flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java

```diff
@@ -0,0 +1,128 @@
+ *
+ * <p>A function that converts various time formats to TIMESTAMP_LTZ type.
+ *
+ * <p>Supported function signatures:
+ *
+ * <ul>
+ *   <li>{@code TO_TIMESTAMP_LTZ(numeric)} -> TIMESTAMP_LTZ(3) <br>
+ *       Converts numeric epoch time in milliseconds to timestamp with local timezone
+ *   <li>{@code TO_TIMESTAMP_LTZ(numeric, precision)} -> TIMESTAMP_LTZ(precision) <br>
+ *       Converts numeric epoch time to timestamp with specified precision (0 as seconds, 3 as
+ *       milliseconds)
+ *   <li>{@code TO_TIMESTAMP_LTZ(timestamp)} -> TIMESTAMP_LTZ(3) <br>
+ *       Parses string timestamp using default format 'yyyy-MM-dd HH:mm:ss'
+ *   <li>{@code TO_TIMESTAMP_LTZ(timestamp, format)} -> TIMESTAMP_LTZ(3) <br>
+ *       Parses string timestamp using input string of format
+ *   <li>{@code TO_TIMESTAMP_LTZ(timestamp, format, timezone)} -> TIMESTAMP_LTZ(3) <br>
+ *       Parses string timestamp using input strings of format and timezone
+ * </ul>
+ *
+ * <p>Example:
+ *
+ * <pre>{@code
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00')}  // Parses string using default format
+ * TO_TIMESTAMP_LTZ(1234567890123)}  // Converts epoch milliseconds
+ * TO_TIMESTAMP_LTZ(1234567890, 0)     // Converts epoch seconds
+ * TO_TIMESTAMP_LTZ(1234567890123, 3)  // Converts epoch milliseconds
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00')  // Parses string using default format
+ * TO_TIMESTAMP_LTZ('2023-01-01T10:00:00', 'yyyy-MM-dd\'T\'HH:mm:ss')} // Parses string using input format
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00', 'yyyy-MM-dd HH:mm:ss', 'UTC')} // Parses string using input format and timezone
+ * }</pre>
+ */
+@Internal
+public class ToTimestampLtzFunction extends BuiltInScalarFunction {
+
+    private static final int DEFAULT_PRECISION = 3;
+
+    public ToTimestampLtzFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.TO_TIMESTAMP_LTZ, context);
+    }
+
+    public @Nullable TimestampData eval(Number epoch, Integer precision) {
+        if (epoch == null || precision == null) {
+            return null;
+        }
+        if (epoch instanceof Float || epoch instanceof Double) {
+            return DateTimeUtils.toTimestampData(epoch.doubleValue(), precision);
+        }
+        return DateTimeUtils.toTimestampData(epoch.longValue(), precision);
+    }
+
+    public @Nullable TimestampData eval(DecimalData epoch, Integer precision) {
+        if (epoch == null || precision == null) {
+            return null;
+        }
+
+        return DateTimeUtils.toTimestampData(epoch, precision);
+    }
+
+    public @Nullable TimestampData eval(Number epoch) {
+        return eval(epoch, DEFAULT_PRECISION);
+    }
+
+    public TimestampData eval(DecimalData epoch) {
+        if (epoch == null) {
+            return null;
+        }
+
+        return DateTimeUtils.toTimestampData(epoch);
+    }
+
+    public @Nullable TimestampData eval(StringData timestamp) {
```

Comment:
Possibly too much for this particular PR - but is it worth considering whether expanding the behaviour to support multiple default format?

For example, using https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatterBuilder.html to support multiple formats? E.g ISO + SQL as default matchers?

I have previously used something like https://www.baeldung.com/java-parsing-dates-many-formats#using-datetimeformatterbuilder when writing UDFs around parsing time to allow for multiple formats within a data source, or to allow for more flexible SQL re-use because I don't have to remember to swap between ISO and SQL or write date formats to cover different zone indcators.

Reply:
Counterpoint - Flink has multiple sources of data, where timestamp strings are not going to be SQL format.

It is frustrating to have to use `UNIX_TIMESTAMP(<date-time>,<format>`) with UTC format string (plus that implementation uses `SimpleDateFormat` so the format string is much more restricted than `DateTimeFormatter`).

`TO_TIMESTAMP` and `TO_TIMESTAMP_LTZ` functions with String input with "by default, SQL or ISO formats are supported" would be valuable.

Unless I've misunderstood as to your concerns about other vendors behaviour?

Reply:
>It is frustrating to have to use UNIX_TIMESTAMP(<date-time>,<format>) with UTC format string (plus that implementation uses SimpleDateFormat so the format string is much more restricted than DateTimeFormatter).

>TO_TIMESTAMP and TO_TIMESTAMP_LTZ functions with String input with "by default, SQL or ISO formats are supported" would be valuable.

some default ISO formats should be supported as support was added long time ago. 
For other topics: looks like it is not related to this PR so feel free either ask a question in ML or create a jira issue for this and submit a PR (if you think it could/should be improved)

Reply:
Thank you, I wasn't sure if it was something to consider as part of these new functions. I will look more into the other default formats and use the appropriate channel.